graduate RetroactiveDefaultStorageClass feature to GA in 1.28

RomanBednar opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
Graduate RetroactiveDefaultStorageClass feature to GA in 1.28.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes # - N/A

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
RetroactiveDefaultStorageClass feature is stable and enabled by default.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/3333
- [Other doc]: https://github.com/kubernetes/enhancements/pull/4021
- [Other doc]: https://github.com/kubernetes/website/pull/41206
```

RomanBednar committed:
graduate RetroactiveDefaultStorageClass feature to GA in 1.28

RomanBednar committed:
test: correct validation test error message

RomanBednar commented:
/retest

carlory commented:
/cc @wojtek-t @smarterclayton

carlory commented:
@RomanBednar As the feature-gate GAed and locked to default,  we need also remove the EnableRetroactiveDefaultStorageClass field from the PersistentVolumeClaimSpecValidationOptions struct. 

https://github.com/kubernetes/kubernetes/pull/111467/files#diff-c713e8919642d873fdf48fe8fb6d43e5cb2f53fd601066ff53580ea655948f0dR2031

carlory commented on pkg/features/kube_features.go:
```diff
@@ -659,7 +659,8 @@ const (
 	// alpha: v1.25
-	//
+	// beta: 1.26
+	// stable: v1.28
```
please left a blank line

RomanBednar replied:
Sure, fixed.

carlory commented:
/test pull-kubernetes-e2e-storage-kind-disruptive

RomanBednar commented:
> @RomanBednar As the feature-gate GAed and locked to default, we need also remove the EnableRetroactiveDefaultStorageClass field from the PersistentVolumeClaimSpecValidationOptions struct.
> 
> https://github.com/kubernetes/kubernetes/pull/111467/files#diff-c713e8919642d873fdf48fe8fb6d43e5cb2f53fd601066ff53580ea655948f0dR2031
> 
> On the other hand, we need remove the feature-gate checked in the pv controller.
> 
> https://github.com/kubernetes/kubernetes/pull/111467/files#diff-9a553e2f939f7ad1470cfe1cf8a48b56b39c6b34fb7ad820c79f2f43602499bdR354

@carlory Thank you for pointing that out. I'm aware of this but thought that a full cleanup of the featuregate should be done one release after GA (so 1.29) to keep a way to opt-out for one more release in case something goes wrong. If this is not desired procedure please let me know, but at this point I don't see any advantage of doing the full cleanup in 1.28.

carlory commented:
At present, I have not found a requirement document that must do a comprehensive cleanup. This seems to be a common practice, for example:

- https://github.com/kubernetes/kubernetes/pull/117718/files
- https://github.com/kubernetes/kubernetes/pull/116741/files
- https://github.com/kubernetes/kubernetes/pull/116233/files


I think it's reasonable, when a feature is stable, to remove the feature-gate control of the code, to ensure that the feature is tested as expected by testing. When the feature-gate is removed in the next version, there will be no additional Code changes, which means there are no other risks

RomanBednar commented:
@carlory According to Kubernetes deprecation policy the feature gate must function for not less than 6 months or 2 releases (whichever is longer) when transitioning from Beta to GA.

https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecation

carlory commented:
@RomanBednar  IMO, we need remove the feature-gate check logic, but remain the feature-gate existing in the command line in order to make the binary work when users specify the feature-gate as same as the default value util the deprecation window is complete.

If the feature-gate isn't locked to default, we need also remain the feature-gate check logic. because users can choose turn off/on a feature.

RomanBednar committed:
test: remove RetroactiveDefaultStorageClass feature gate

Since the feature is GA and locked to true, tests can no longer set it
to false. Cleaning up by removing all references to this feature gate
from tests.

Feature gate will be removed in v1.29.

RomanBednar committed:
remove RetroactiveDefaultStorageClass feature gate checks

jsafrane commented:
/lgtm
/approve

msau42 commented:
/lgtm
/approve
